[FIX] website_sale_stock: fix max combo quantity

ltinel opened the pull request:
Before this fix, it was possible to add more combo products than available to the cart. Indeed, the available quantity was only checked in the combo configurator, but not in other places where the quantity could be updated (such as the product page or the cart).

This fix introduces a `max_combo_quantity` concept which is based on the available quantities of its items.

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/188860.png)](https://mergebot.odoo.com/odoo/odoo/pull/188860)

chevalierv commented on addons/website_sale_stock/static/src/js/variant_mixin.js:
```diff
@@ -61,6 +61,14 @@ VariantMixin._onChangeCombinationStock = function (ev, $parent, combination) {
+            ctaWrapper.classList.replace('d-flex', 'd-none');
+            ctaWrapper.classList.add('out_of_stock');
+        }
+        // TODO(loti): show "out of stock" message?
```
👀

ltinel replied:
I'll have a chat with Jean to discuss what we should do here :)

ltinel commented on addons/website_sale_stock/models/product_product.py:
```diff
@@ -15,15 +14,20 @@ def _has_stock_notification(self, partner):
-        if not self.allow_out_of_stock_order:
-            website = website or self.env['website'].get_current_website()
-            # When the cron is run manually, request has no attribute website, and that would cause a crash
-            # so we check for it
```
For the reviewer: the cron in question (`ir_cron_send_availability_email`) doesn't rely on this method anymore, so no need for the check.

ltinel commented on addons/website_sale_stock/models/product_product.py:
```diff
@@ -15,15 +14,20 @@ def _has_stock_notification(self, partner):
         return partner in self.stock_notification_partner_ids
 
     def _get_cart_qty(self, website=None):
-        if not self.allow_out_of_stock_order:
```
@reviewers: I had a look at the code and we don't actually need this condition. In fact, this condition makes the code a bit inconsistent, so I removed it

chevalierv reviewed (changes requested):
Hello @ltinel 

Thanks for addressing the issues about max quantity with combos. I feel like this is a huge change for stable version. Please double check if it's really needed with PO.

See my comments for other issues.

Have a nice day!

ltinel commented:
Note: the latest force-push was just a rebase (commit unchanged).

ltinel commented on addons/website_sale_stock/static/src/xml/website_sale_stock_product_availability.xml:
```diff
@@ -2,7 +2,14 @@
 
 <templates>
     <t t-name="website_sale_stock.product_availability">
-        <div t-if="is_storable and !prevent_zero_price_sale" id="product_stock_availability">
+        <div t-if="!prevent_zero_price_sale" id="product_stock_availability">
+            <t t-call="website_sale_stock.storable_product_availability"/>
+            <t t-call="website_sale_stock.combo_product_availability"/>
+        </div>
+    </t>
```
This change is not essential IMO (it just shows a `Not available` message if a combo product isn't available), so I'll keep it for master :)

ltinel commented:
> Hello @ltinel
> 
> Thanks for addressing the issues about max quantity with combos. I feel like this is a huge change for stable version. Please double check if it's really needed with PO.
> 
> See my comments for other issues.
> 
> Have a nice day!

@chevalierv in the meantime, a customer reported this issue (https://www.odoo.com/odoo/project.task/4614309), so I guess we should fix it in stable? The inventory team created a fix PR (https://github.com/odoo/odoo/pull/202316) which is way smaller than mine, but it's insufficient IMO.

@Jean-boje WDYT?

chevalierv reviewed (approved):
Ok to fix this in stable since you removed the cosmetics changes and there is a ticket for that! 

robodoo delegate+

ltinel commented on addons/website_sale/controllers/main.py:
```diff
@@ -864,17 +864,37 @@ def cart_update_json(
             no_variant_attribute_value_ids=no_variant_attribute_value_ids,
             **kwargs
         )
+
```
This one serves as separation between the main line update and the linked line updates (to which I've added quite a few lines) so I'd like to keep it if you don't mind :)

ltinel commented:
@robodoo r+

ltinel commented:
@robodoo r+
@robodoo rebase-ff

robodoo commented:
Merge method set to rebase and fast-forward.

robodoo commented:
@ltinel 'ci/l10n' failed on this reviewed PR.

robodoo commented:
@ltinel 'ci/runbot' failed on this reviewed PR.

ltinel commented:
@robodoo r+
